nisystemlink/cross-service/GenericApi.py

```diff
@@ -0,0 +1,1040 @@
+            local_http_conn.request(
+                "POST",
+                "/nidataframe/v1/tables/{}/data".format(frame_id),
+                body=json.dumps(body),
+                headers=self._headers,
+            )
+            r = local_http_conn.getresponse()
+            if r.status == 204:
+                upload_success = True
+            else:
+                num_trys += 1
+                print(r.status)
+                print("Upload attempt {} failed".format(num_trys))
+        local_http_conn.close()
+        return upload_success
+
+    async def get_test_id_from_file_id(self, file_id: str):
+        """Get the test id that corresponds to a given file."""
+        response = fis_models.QueryResponse(
+            await self._fis_handle.list_available_files_get(id=file_id)
+        )
+        files = response.links.available_files
+        return files[0].properties["test_id"]
+
+    async def create_datatable(
+        self,
+        column_schema: Dict[str, Any],
+        datatable_type: DatatableType = DatatableType.RAW,
+        workspace: str = None,
+    ) -> str:
+        """Create a datatable."""
+        datatable = ""
+        if workspace:
+            column_schema["workspace"] = workspace
+        body = json.dumps(column_schema)
+        self._http_conn.request("POST", "/nidataframe/v1/tables", body=body, headers=self._headers)
+        response = self._http_conn.getresponse()
+        if response.status >= 400:
+            raise InternalError(response.read())
+        else:
+            payload = ""
+            while chunk := response.read(200):
+                payload += repr(chunk)
+            frames = json.loads(payload.replace("'", '"')[2:-1])
+            datatable = frames["id"]
+        body = json.dumps({"properties": {"datatableType": datatable_type}})
+        self._http_conn.request(
+            "PATCH", f"/nidataframe/v1/tables/{datatable}", body=body, headers=self._headers
+        )
+        response = self._http_conn.getresponse()
+        if response.status >= 400:
+            raise InternalError(response.read())
+        else:
+            response.read()
+        return datatable
+
+    async def get_workspace_id(self, desired_workspace: str) -> str:
+        """Get the id of the specified workspace."""
+        desired_workspace = urllib.parse.quote(desired_workspace)
+        self._http_conn.request(
+            "GET", f"/niuser/v1/workspaces?name={desired_workspace}", headers=self._headers
+        )
+        response = self._http_conn.getresponse()
+        if response.status >= 400:
+            raise InternalError(response.read())
+        payload = ""
+        while chunk := response.read(200):
+            payload += repr(chunk)
+        query_response = json.loads(payload.replace("'", '"')[2:-1])
+        workspace_id = query_response["workspaces"][0]["id"]
+        return workspace_id
+
+    async def get_specific_cycle_table_id(
```

Comment:
This and the methods below are all based on the existence of "proprietary" custom properties, so they should be in a separate package